### PR TITLE
[ready-for-core-team-review]CRM-16189 Added settings for 'Always post to Accounts Receivable'

### PR DESCRIPTION
**Pull request by pradpnayak**
---
- CRM-16189: Improve support for Accrual Method bookkeeping
  https://issues.civicrm.org/jira/browse/CRM-16189

### CODE REVIEW

File: CRM/Contribute/BAO/Contribution.php [hunk 1]
```diff
@@ -220,6 +220,9 @@ public static function add(&$params, $ids = array()) {
 
     //add Account details
     $params['contribution'] = $contribution;
+    if (empty($contributionID)) {
+      self::recordAlwaysAccountsReceivable($params);
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: Let's try to be more consistent in names to help future developers. If the setting is always_post_to_accounts_receivable let's make this the same.

File: CRM/Contribute/BAO/Contribution.php [hunk 2]
```diff
@@ -5245,4 +5248,44 @@ protected function assignTestValue($fieldName, &$fieldDef, $counter) {
...
 
+  /**
+   * Create Accounts Receivable financial trxn entry for Completed Contribution.
+   *
+   * @param array $params
+   *
+   * @param string $context
+   *
+   */
+  public static function recordAlwaysAccountsReceivable(&$params, $context = NULL) {
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: Rename to alwaysPostToAccountsReceivable

File: CRM/Contribute/BAO/Contribution.php [hunk 3]
```diff
@@ -5245,4 +5248,44 @@ protected function assignTestValue($fieldName, &$fieldDef, $counter) {
...
+    if (!self::checkContributeSettings('always_post_to_accounts_receivable')) {
+      return NULL;
+    }
+    $statusId = $params['contribution']->contribution_status_id;
+    $contributionStatuses = CRM_Contribute_PseudoConstant::contributionStatus(NULL, 'name');
+    $contributionStatus = empty($statusId) ? NULL : $contributionStatuses[$statusId];
+    // Return if contribution status is not completed.
+    if ($contributionStatus != 'Completed') {
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: I'm not sure this is correct. Let's discuss in scrum.

File: CRM/Contribute/BAO/Contribution.php [hunk 4]
```diff
@@ -5245,4 +5248,44 @@ protected function assignTestValue($fieldName, &$fieldDef, $counter) {
     }
   }
 
+  /**
+   * Create Accounts Receivable financial trxn entry for Completed Contribution.
+   *
+   * @param array $params
+   *
+   * @param string $context
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: Could you provide a brief explanation of what value(s) $context might be expected to hold?

File: CRM/Contribute/BAO/Contribution.php [hunk 5]
```diff
@@ -5245,4 +5248,44 @@ protected function assignTestValue($fieldName, &$fieldDef, $counter) {
     }
   }
 
+  /**
+   * Create Accounts Receivable financial trxn entry for Completed Contribution.
+   *
+   * @param array $params
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: Are these the $params used to create/seach for a contribution? A bit of explanation would help.

File: settings/Contribute.setting.php [hunk 6]
```diff
@@ -165,4 +165,19 @@
...
+  'always_post_to_accounts_receivable' => array(
+    'group_name' => 'Contribute Preferences',
+    'group' => 'contribute',
+    'name' => 'always_post_to_accounts_receivable',
+    'type' => 'Integer',
+    'html_type' => 'checkbox',
+    'quick_form_type' => 'Element',
+    'default' => 0,
+    'add' => '4.7',
+    'title' => 'Always post to Accounts Receivable',
```
Comment (JoeMurray) [inline settings/Contribute.setting.php]: Add a question mark: 'Always post to Accounts Receivable?'

**Comment by monishdeb** [comment]
@pradpnayak can you please rebase your PR ?

**Comment by pradpnayak** [comment]
@monishdeb rebased PR

**Comment by jitendrapurohit** [comment]
@pradpnayak Test failures seems related. Can you please have a look at them ?

**Review by jitendrapurohit — APPROVED** [review]
Tested this on Contributions, Events and Membership with pending and completed status. Financial Trxn entries are correctly updated for `account_receivable` financial account.

**Comment by jitendrapurohit** [comment]
Test failure `CRM_Contribute_Form_ContributionTest::testSubmitCreditCardPayPal` is pre-existing/unrelated.

For `CRM_Dedupe_MergerTest::testGetOrganizationMatches` - it works fine on my local after applying the patch.

**Pull request merged**